core/src/main/scala/kafka/log/AbstractIndex.scala

```diff
@@ -120,45 +145,84 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
       }
 
       /* memory-map the file */
-      _length = raf.length()
+      _length = Some(raf.length())
       val idx = {
         if (writable)
-          raf.getChannel.map(FileChannel.MapMode.READ_WRITE, 0, _length)
+          raf.getChannel.map(FileChannel.MapMode.READ_WRITE, 0, _length.get)
         else
-          raf.getChannel.map(FileChannel.MapMode.READ_ONLY, 0, _length)
+          raf.getChannel.map(FileChannel.MapMode.READ_ONLY, 0, _length.get)
       }
       /* set the position in the index for the next entry */
       if(newlyCreated)
         idx.position(0)
       else
-        // if this is a pre-existing index, assume it is valid and set position to last entry
+      // if this is a pre-existing index, assume it is valid and set position to last entry
         idx.position(roundDownToExactMultiple(idx.limit(), entrySize))
-      idx
+      _mmap = idx
     } finally {
+      initialized = true
       CoreUtils.swallow(raf.close(), this)
     }
   }
+  
+  protected def mmap: MappedByteBuffer = {
+    if (!initialized) {
+      inLock(lock) {
+        if (!initialized) {
+          initializeMmapAndLength
+        }
+      }
+    }
+    _mmap
+  }
 
   /**
    * The maximum number of entries this index can hold
    */
   @volatile
-  private[this] var _maxEntries = mmap.limit() / entrySize
+  private[this] var _maxEntries: Option[Int] = None
 
   /** The number of entries in this index */
   @volatile
-  protected var _entries = mmap.position() / entrySize
+  protected var _entries: Option[Int] = None
 
   /**
    * True iff there are no more slots available in this index
    */
-  def isFull: Boolean = _entries >= _maxEntries
-
-  def maxEntries: Int = _maxEntries
+  def isFull: Boolean = entries >= maxEntries
+
+  // This can trigger IO operation.
+  def maxEntries: Int = {
+    if (_maxEntries.isEmpty) {
+      inLock(lock) {
+        if (_maxEntries.isEmpty)
+          _maxEntries = Some(mmap.limit() / entrySize)
+      }
+    }
+    _maxEntries.get
+  }
 
-  def entries: Int = _entries
+  // This can trigger IO operation.
+  def entries: Int = {
+    if (_entries.isEmpty) {
+      inLock(lock) {
+        if (_entries.isEmpty)
+          _entries = Some(mmap.position() / entrySize)
+      }
+    }
+    _entries.get
+  }
 
-  def length: Long = _length
+  // This can trigger IO operation.
+  def length: Long = {
+    if (!initialized) {
+      inLock(lock) {
+        if (!initialized)
+          initializeMmapAndLength
+      }
+    }
```

Comment:
Because this block is repeated a few times, could we fold it into a method?

Reply:
Done.